apps/cave/components/AMM/Swap/CustomRecipient.tsx

```diff
@@ -11,21 +11,33 @@ import {
   useMultiStyleConfig,
 } from '@concave/ui'
 import { isAddress } from 'ethers/lib/utils'
-import { useReducer, useRef, useState } from 'react'
-import { useEnsAddress } from 'wagmi'
+import { useReducer, useRef } from 'react'
+import { usePrevious } from 'react-use'
+import { useDebounce } from 'usehooks-ts'
+import { Address, useEnsAddress } from 'wagmi'
 
-export const CustomRecipient = ({ onChangeRecipient }) => {
+export const CustomRecipient = ({
+  onChangeRecipient,
+}: {
+  onChangeRecipient: (r?: Address) => void
+}) => {
   const [isOpen, toggle] = useReducer((s) => !s, false)
 
-  const [recipient, setRecipient] = useState('')
+  const [recipient, setRecipient] = useReducer((_, r) => {
+    // if user types an address, set it as recipient
+    if (isAddress(r)) onChangeRecipient(r)
+    return r
+  }, '')
+  const debouncedRecipient = useDebounce(recipient, 500)
 
   const { data: ensAddress, isLoading } = useEnsAddress({
-    name: recipient,
-    enabled: !isAddress(recipient),
-    onSuccess: (address) => {
-      onChangeRecipient(address)
-    },
+    name: debouncedRecipient,
+    enabled: !isAddress(debouncedRecipient),
   })
+  // if user types an ens name,
+  // check if it resolved to an address and is not the same as the previous render (to avoid infinite loops)
+  const prevEnsAddress = usePrevious(ensAddress)
+  if (prevEnsAddress !== ensAddress && isAddress(ensAddress)) onChangeRecipient(ensAddress)
```

Comment:
kinda like the way this component self contains this logic, this is a needed workaround, the other option would be to lift the state up to it's parent component
(maybe a `useRecepient` hook called on parent could work?) that's an idea I just had

Reply:
implemented hook idea, way better